### PR TITLE
add prop iterator

### PR DESCRIPTION
**Pull request by nevermore3**

### CODE REVIEW

File: src/context/test/IteratorTest.cpp [hunk 1]
```diff
@@ -746,6 +746,109 @@ TEST(IteratorTest, Join) {
...
+            result.emplace_back(std::move(v));
+        }
+        EXPECT_EQ(result.size(), 10);
+        EXPECT_EQ(result, expected);
+    }
+}
+
+TEST(IteratorTest, EdgeProp) {
+    DataSet ds;
+    ds.colNames = {kSrc, kType, kRank, kDst, "like:prop1", "like:prop2"};
```
Comment (Shylock-Hg) [inline src/context/test/IteratorTest.cpp:805]: The `kSrc`, `kType` , `kRank` and `kDst` won't always return.

File: src/context/Iterator.cpp [hunk 2]
```diff
@@ -388,6 +388,208 @@ size_t JoinIter::buildIndexFromJoinIter(const JoinIter* iter, size_t segIdx) {
...
+    if (!vidVal.isStr()) {
+        return Value::kNullValue;
+    }
+    Vertex vertex;
+    vertex.vid = vidVal.getStr();
+    auto& tagPropsMap = dsIndex_.propsMap;
+    bool isVertexProps = true;
+    for (auto& tagProp : tagPropsMap) {
+        auto& row = *(iter_->row_);
+        auto& colIndexs = tagProp.second.colIndexs;
```
Comment (laura-ding) [inline src/context/Iterator.cpp:491]: I think the `colIndexs` is useless. propIndices has included it.

File: src/context/Iterator.cpp [hunk 3]
```diff
@@ -388,6 +388,208 @@ size_t JoinIter::buildIndexFromJoinIter(const JoinIter* iter, size_t segIdx) {
...
+Value PropIter::getEdge() const {
+    if (!valid()) {
+        return Value::kNullValue;
+    }
+    auto row = *(iter_->row_);
+    Edge edge;
+    auto& edgePropsMap = dsIndex_.propsMap;
+    bool isEdgeProps = true;
+    for (auto& edgeProp : edgePropsMap) {
+        for (auto& index : edgeProp.second.colIndexs) {
```
Comment (laura-ding) [inline src/context/Iterator.cpp:522]: ditto

**Review by CPWstatic — APPROVED** [review]

File: src/context/Iterator.cpp (hunk 2 above)
Comment (nevermore3) [reply src/context/Iterator.cpp]: done

File: src/context/Iterator.cpp (hunk 3 above)
Comment (nevermore3) [reply src/context/Iterator.cpp]: done

**Review by laura-ding — APPROVED** [review]
Pay attention to somewhere need to add a check for the new iterator type.

File: src/context/Iterator.cpp [hunk 4]
```diff
@@ -388,6 +388,204 @@ size_t JoinIter::buildIndexFromJoinIter(const JoinIter* iter, size_t segIdx) {
...
+const Value& PropIter::getColumn(const std::string& col) const {
+    if (!valid()) {
+        return Value::kNullValue;
+    }
+
+    auto& logicalRow = *iter_;
+    auto index = dsIndex_.colIndices.find(col);
+    if (index == dsIndex_.colIndices.end()) {
+        return Value::kNullValue;
+    } else {
```
Comment (monadbobo) [inline src/context/Iterator.cpp:446]: No need for an else here.

Comment (nevermore3) [reply src/context/Iterator.cpp:446]: make senses

**Review by laura-ding — APPROVED** [review]

**Review by dutor — APPROVED** [review]

**Pull request merged**